### PR TITLE
LUKS OPAL support

### PR DESCRIPTION
**Pull request by vojtechtrefny**
Support for OPAL self-encrypting drives using the latest cryptsetup feature that allows combining LUKS and SED OPAL drives.

Marking as a draft for now, I don't plan to do much changes but I might change some smaller things based on the blivet and anaconda support.

### CODE REVIEW

**Comment by vojtechtrefny** [comment]
Jenkins, test this please.

File: src/plugins/crypto.c [hunk 1]
```diff
@@ -3504,3 +3621,190 @@ gboolean bd_crypto_fvault2_close (const gchar *fvault2_device, GError **error) {
...
+        return FALSE;
+    }
+
+    ret = ioctl (fd, IOC_OPAL_GET_STATUS, &st);
+    if (ret < 0) {
+        g_set_error (error, BD_CRYPTO_ERROR, BD_CRYPTO_ERROR_DEVICE,
+                     "Failed to get opal status for the device '%s': %s",
+                     device,
+                     strerror_l (-ret, c_locale));
+        return FALSE;
```
Comment (tbzatek) [inline src/plugins/crypto.c:3654]: leaking `fd`? Twice.

File: src/plugins/crypto.c [hunk 2]
```diff
@@ -3504,3 +3621,190 @@ gboolean bd_crypto_fvault2_close (const gchar *fvault2_device, GError **error) {
...
+    }
+
+    ret = crypt_get_hw_encryption_type (cd);
+	if (ret != CRYPT_OPAL_HW_ONLY && ret != CRYPT_SW_AND_OPAL_HW) {
+		g_set_error (&l_error, BD_CRYPTO_ERROR, BD_CRYPTO_ERROR_DEVICE,
+                     "Device %s isn't a LUKS HW-OPAL device.", device);
+        bd_utils_report_finished (progress_id, l_error->message);
+        g_propagate_error (error, l_error);
+        crypt_free (cd);
+        return FALSE;
```
Comment (tbzatek) [inline src/plugins/crypto.c:3720]: weird indentation

**Review by tbzatek — COMMENTED** [review]
Yeah the API looks fine with me

File: src/plugins/crypto.c (hunk 1 above)
Comment (vojtechtrefny) [reply src/plugins/crypto.c:3670]: Fixed.

File: src/plugins/crypto.c (hunk 2 above)
Comment (vojtechtrefny) [reply src/plugins/crypto.c]: Who put tabs in my code? :-)

File: src/plugins/crypto.c [hunk 3]
```diff
@@ -36,6 +36,8 @@
 #include <volume_key/libvolume_key.h>
 #endif
 
+#include <linux/sed-opal.h>
```
[the remark this answers is not in the document]
Comment (vojtechtrefny) [reply src/plugins/crypto.c:40]: I planned to add check for this and forgot. There probably aren't many systems with cryptsetup 2.7 and kernel <= 6.4 (or with OPAL disabled), but it's better to check this too, fixed.

**Comment by vojtechtrefny** [comment]
> Marking as a draft for now, I don't plan to do much changes but I might change some smaller things based on the blivet and anaconda support.

I have the blivet support finished and tested so I think there won't be any more changes here, so marking this as ready for review.

**Review by tbzatek — APPROVED** [review]
One last detail, otherwise good to go!

File: src/plugins/crypto.c [hunk 4]
```diff
@@ -438,6 +445,19 @@ gboolean bd_crypto_is_tech_avail (BDCryptoTech tech, guint64 mode, GError **erro
                 return FALSE;
             } else
                 return TRUE;
+        case BD_CRYPTO_TECH_SED_OPAL:
+#if !defined(LIBCRYPTSETUP_27) && !defined(HAVE_LINUX_OPAL)
```
[the remark this answers is not in the document]
Comment (vojtechtrefny) [reply src/plugins/crypto.c]: Yes, thanks. :facepalm:

**Pull request merged**